### DISCUSSION TITLE
Custom URL based API

### DISCUSSION THREAD

**Discussion by jeevanantham123** [vercel/next.js#25894] [Ideas]
Hi,
I am new to Next, hereby asking out of curiosity. Is there any way to have custom routes for API, instead of creating a new file that will match the URL?

The problem here is if I want to create a CRUD request for a post:
Create - post request ( pages/api/post/index.js)
Read - get request ( pages/api/post/index.js)
(I can combine those above two  API's in a single file)
Update - post request ( pages/api/post/updatePost.js)
(I want to create a new file to differentiate from the previous Post API)
Delete - delete request (pages/api/post/updatePost.js)
I can use the same updatePost file to use the delete function as it is a delete request coz I can differentiate the function using request type.

As you can see,
/api/post
-- index.js
----create
----read
--updatePost.js
----update
----delete

If I care about API pathName I must create a new deletePost file to take care of the delete request, coz updatePost doesn't match with delete functionality.

So, I suggest there must be an easiest way to create custom domains match functions, without creating multiple file architectures.
we need the liberty to match the route with exact functions.

example:
post/index.js
---router.match('/add')
---router.match('/read')
---router.match('/update')
---router.match('/delete')

Here u can see, one file is enough to handle CRUD operations.

**Comment by spro** [comment]
There are two good ways to do this... to exactly match your stated API you would want to use [Dynamic API Routes](https://nextjs.org/docs/api-routes/dynamic-api-routes) - specifically your handler would live at `pages/api/post/[action].js` and you could choose your action based on that `action` parameter.

That said a potentially more standard / RESTful approach is to design your API this way (also taking advantage of dynamic routes for the second file):

`pages/api/posts.js` would handle:

* `GET api/posts` (list)
* `POST api/posts` (create)

`pages/api/posts/[id].js` would handle:

* `GET api/posts/123` (read)
* `PUT api/posts/123` (update)
* `DELETE api/posts/123` (delete)

**Reply by jeevanantham123** [reply]
Thanks for the reply, the thing I searched for is [..slug.js]
coz, what if I want to add an API that performs the following operations,

- getPostbyId
- deletePostbyId
- updatePostbyId
- addPostbyUserId
along with CRUD.

these are some use cases that will be implemented in an application,

here I need to use a slug and differentiate the function matching by params. 
Right?

**Reply by spro** [reply]
I personally would use the HTTP method within the handler to distinguish actions, e.g. in `api/things/[id].js`

```javascript
export default async function handler(req, res) {
    const thing_id = req.query.id
    if (req.method == 'GET') {
        res.json(await backend.things.getThing(thing_id))
    } else if (req.method == 'PUT') {
        res.json(await backend.things.updateThing(thing_id, req.body))
    }
   ... delete, etc.
}
```

However if you want explicit actions as route params like `/things/2/update` you can name your directory with brackets also, e.g. in `pages/api/things/[id]/[action].ts`

```javascript
export default function handler(req, res) {
    const thing_id = req.query.id
    const action = req.query.action
    if (action == 'update') updateThing(thing_id) ...
}
```
